cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/EOGWorklist.kt

```diff
@@ -0,0 +1,301 @@
+        private set
+
+    /** A list of all nodes which have already been visited. */
+    private val alreadySeen = mutableListOf<K>()
+
+    constructor(globalState: MutableMap<K, State<N, V>> = mutableMapOf()) : this() {
+        this.globalState = globalState
+    }
+
+    /**
+     * The actual worklist, i.e., elements which still have to be analyzed and the state which
+     * should be considered there.
+     */
+    private val nodeOrder: MutableList<Pair<K, State<N, V>>> = mutableListOf()
+
+    /**
+     * Adds [newNode] and the [state] to the [globalState] (i.e., computes the [State.lub] of the
+     * current state there and [state]). Returns true if there was an update.
+     */
+    fun update(newNode: K, state: State<N, V>): Boolean {
+        val (newGlobalState, update) = globalState[newNode]?.lub(state) ?: Pair(state, true)
+        if (update) {
+            globalState[newNode] = newGlobalState
+        }
+        return update
+    }
+
+    /**
+     * Pushes [newNode] and the [state] to the worklist or updates the currently available entry for
+     * the node. Returns `true` if there was a change which means that the node has to be analyzed.
+     * If it returns `false`, the [newNode] wasn't added to the worklist as the state didn't change.
+     */
+    fun push(newNode: K, state: State<N, V>): Boolean {
+        val currentEntry = nodeOrder.find { it.first == newNode }
+        val update: Boolean
+        val newEntry =
+            if (currentEntry != null) {
+                val (newState, update2) = currentEntry.second.lub(state)
+                update = update2
+                if (update) {
+                    nodeOrder.remove(currentEntry)
+                }
+                Pair(currentEntry.first, newState)
+            } else {
+                update = true
+                Pair(newNode, state)
+            }
+        if (update) nodeOrder.add(newEntry)
+        return update
+    }
+
+    /** Determines if there are still elements to analyze */
+    fun isNotEmpty() = nodeOrder.isNotEmpty()
+    /** Determines if there are no more elements to analyze */
+    fun isEmpty() = nodeOrder.isEmpty()
+
+    /** Removes a [Node] from the worklist and returns the [Node] together with its [State] */
+    fun pop(): Pair<K, State<N, V>> {
+        val node = nodeOrder.removeFirst()
+        alreadySeen.add(node.first)
+        return node
+    }
+
+    /** Checks if [currentNode] has already been visited before. */
+    fun hasAlreadySeen(currentNode: K) = currentNode in alreadySeen
+
+    /** Computes the meet over paths for all the states in [globalState]. */
+    fun mop(): State<N, V> {
+        val firstKey = globalState.keys.firstOrNull()
+        val state = globalState[firstKey]
+        for ((_, v) in globalState) {
+            state?.lub(v)
+        }
+
+        return state ?: State()
+    }
+}
+
+/** A very simple implementation of the worklist algorithm. */
+class EOGWorklist {
```

Comment:
a lot of classes.. why is EOGworklist not deriving from a "wordlist", but somehow has it in the transformation as an argument? Maybe different naming?

Reply:
You probably could just get rid of this class altogether and directly have those two functions as top-level functions.